js/classes/entities/foreground/weather_system.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * WeatherSystem manages what weather to spawn into a zone.
+ * 
+ * @author Caleb Krauter
+ */
+class WeatherSystem {
+
+    /**
+     * Pass in the type of weather, intensity and time of day.
+     * Direction of rain or snow is random.
+     * Rain or snow will always include clouds.
+     * Intensity is used for both rain/snow and clouds.
+     * 
+     * @param {String} type warm, hot, cloudy, rain, snow.
+     * @param {int} intensity a value between 0 and 5.
+     * @param {String} time day or night.
+     */
+    constructor(type, intensity, time) {
+        this.type = type;
+        this.intensity = intensity;
+        this.time = time;
+
+        // Rainy day or night
+        this.makeHeavens();
+
+        if (this.type === "rain") {
+            this.makeClouds();
+            this.makePrecipitation();
+        }
+        else if (this.type === "snow") {
+            this.makeClouds();
+            this.makePrecipitation();
+            this.addSurfaceSnow();
+        } else if (this.type === "cloudy") {
+            this.makeClouds();
+            // Default to sunny with white clouds
+        }
+        // Defaults to sunny/(night with moon) and no clouds.
+        else {
+            this.makeHeavens();
+        }
+    }
+
+    /**
+     * Adds a sun/moon and updates the sky color.
+     */
+    makeHeavens() {
+        if (this.time === "night") {
+            BG_COLOR = COLORS.SKY_DARK_GREY;
+            GAME.addEntity(new Sun(new Vector(Camera.SIZE.x - 2 * Sun.SCALED_SIZE, Sun.SCALED_SIZE - 100), Sun.MOON), -1);
+
+        } else {
+            switch (this.type) {
+                case "warm":
+                    BG_COLOR = COLORS.SKY_BLUE;
+                    GAME.addEntity(new Sun(new Vector(Camera.SIZE.x - 2 * Sun.SCALED_SIZE, Sun.SCALED_SIZE - 100), Sun.VILLAGE), -1);
+                    break;
+                case "hot":
+                    BG_COLOR = COLORS.SKY_HOT_SKY;
+                    GAME.addEntity(new Sun(new Vector(Camera.SIZE.x - 2 * Sun.SCALED_SIZE, Sun.SCALED_SIZE - 100), Sun.LAVA), -1);
+                    break;
+                case "cloudy":
+                    BG_COLOR = COLORS.SKY_GREY;
+                    break;
+                case "rain":
+                    BG_COLOR = COLORS.SKY_GREY;
+                    break;
+                case "snow":
+                    BG_COLOR = COLORS.SKY_SNOW_GREY;
+                    break;
+                default:
+                    break;
+            }
+        }
+
+    }
+
+    /**
+     * Adds clouds.
+     */
+    makeClouds() {
+        // Make sure to only have a number of clouds that is divisble by 100. For example, 15 is bad value but 25 is a good value.
+        const numOfClouds = [2, 5, 10, 20, 25, 50];
```

Comment:
Good call composing the various numbers of clouds in an array.  It makes the randomization much cleaner.